src/symbol_table.rs

```diff
@@ -1,14 +1,118 @@
+use std::borrow::Borrow;
 use std::collections::HashMap;
+use std::hash::{Hash, Hasher};
+use std::ops::Deref;
+use std::rc::Rc;
 
 use crate::constants::v1_0;
 use crate::types::SymbolId;
 
+/// Stores or points to the text of a given [Symbol].
+#[derive(Debug, Eq)]
+enum SymbolText {
+    // This Symbol refers to a string in the symbol table
+    Shared(Rc<str>),
+    // This Symbol owns its own text
+    Owned(String),
+}
+
+impl AsRef<str> for SymbolText {
+    fn as_ref(&self) -> &str {
+        match self {
+            SymbolText::Owned(ref text) => text.as_str(),
+            SymbolText::Shared(ref rc) => rc.as_ref(),
+        }
+    }
+}
+
+impl<A: AsRef<str>> PartialEq<A> for SymbolText {
+    fn eq(&self, other: &A) -> bool {
+        // Compare the Symbols' text, not their ownership models
+        self.as_ref() == other.as_ref()
+    }
+}
+
+impl Hash for SymbolText {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        // Hash the Symbol's text, ignore where/how it's stored.
+        self.as_ref().hash(state)
+    }
+}
+
+impl Clone for SymbolText {
+    fn clone(&self) -> Self {
+        match self {
+            SymbolText::Owned(text) => SymbolText::Owned(text.to_owned()),
+            SymbolText::Shared(text) => SymbolText::Shared(Rc::clone(text)),
+        }
+    }
+}
+
+// TODO: Doc comment. Fully resolved symbol that can be handed out. Required to have text.
+#[derive(Debug, Hash, Clone, Eq)]
+pub struct Symbol {
+    text: SymbolText, // TODO: As an application-level entity, does this need to store an Option<SymbolId>?
+}
+
+impl<A: AsRef<str>> PartialEq<A> for Symbol {
+    fn eq(&self, other: &A) -> bool {
+        self.text.as_ref() == other.as_ref()
+    }
+}
+
+impl PartialEq<Symbol> for String {
+    fn eq(&self, other: &Symbol) -> bool {
+        self.as_str() == other.as_ref()
+    }
+}
+
+impl PartialEq<Symbol> for &str {
+    fn eq(&self, other: &Symbol) -> bool {
+        self == &other.as_ref()
+    }
+}
```

Comment:
PR tour: these impls allow `Symbol` instances to be compared to other text types in (e.g.) an `assert_eq!`, which streamlines a lot of unit tests.